Sources/Plasma/PubUtilLib/plAvatar/plPhysicalControllerCore.cpp

```diff
@@ -375,194 +375,110 @@ plWalkingStrategy::plWalkingStrategy(plAGApplicator* rootApp, plPhysicalControll
+    }
+    velocity.fZ = std::max(velocity.fZ, kGravity);
+
+    // Some extra oomph to keep us on the ground.
+    if (IsOnGround())
+        velocity.fZ *= 1.1f;
+
+    if (!IsOnGround() && std::fabs(velocity.fX) < 0.001f && std::fabs(velocity.fY) < 0.001f) {
+        // If we're airborne and the velocity isn't set, use the velocity from
+        // the last frame so we maintain momentum.
+        velocity.fX = achievedVelocity.fX;
+        velocity.fY = achievedVelocity.fY;
+    } else if (IsOnGround()) {
+        // Project velocity by the ground normal to prevent speeding up
+        // when running uphill (grrr...) Note that if the ground is a proxy
+        // collider, you may have collisions from both the wall (z=0.0) and
+        // the floor (z=-1.0)... be sure to use the floor.
+        plKey groundPhys = fController->GetGround();
+        const hsVector3* groundNormal = nullptr;
+        for (const auto& contact : fContacts) {
+            if (contact.PhysHit == groundPhys) {
+                if (!groundNormal || std::abs(contact.Normal.fZ) > std::abs(groundNormal->fZ))
+                    groundNormal = &contact.Normal;
             }
-        } else if (std::fabs(velocity.fZ) < 0.001f) {
-            float zcomp = 1.f - std::max(0.f, ground->Normal.fZ);
-            velocity.fZ = achievedVelocity.fZ + (kGravity * zcomp * delSecs);
         }
 
-        // Kill upward velocity if we're just running along the ground.
-        if (ground && !IsControlledFlight())
-            velocity.fZ = std::min(0.f, velocity.fZ);
+        if (groundNormal) {
+            // The Y component is forward, so use it always. Further, we can go up some pretty
+            // doggone steep slopes, resulting in the theoretical variantion of half of our velocity.
+            // Debounce that by capping the increase or decrease.
+            const float yFactor = (achievedVelocity.fZ < 0.f) ?
+                // Going downhill leg - note that even a small increase in velocity is VERY obvious here.
+                // Therefore, we make no adjustment.
+                1.f :
+                // Going uphill leg. Ideally this would be some kind of non-linear calculation.
+                // That way, going up stairs is much more noticable than running uphill in Relto.
+                // But MEH, you do it.
+                std::max(1.f - std::abs(groundNormal->fY), .9f);
+            velocity.fX *= yFactor;
+            velocity.fY *= yFactor;
+        }
     }
 
-    // Limit final requested downward velocity to the magnitude of the acceleration of gravity.
-    // It's not the way R/L works, but it yields a better visual result.
-    if (!(fFlags & kGroundContact))
-        velocity.fZ = std::max(velocity.fZ, kGravity);
-
-    // Reset vars and move the controller
-    fController->SetPushingPhysical(nullptr);
-    fController->SetFacingPushingPhysical(false);
-    fContacts.clear();
+    // Now Cyan's good old "slide along all the contacts" code.
```

Comment:
Yeah, uh, the best I can tell you is that this some code from Cyan that's been hanging along for the ride. `fContacts` is indeed a collection of every contact point that the avatar capsule is in contact with. The goal is to project the avatar's velocity along all of those contacts so that, for example, when you clip a wall, you wind up sliding along that wall instead of simply stopping dead. I'm honestly not super sure how the math works except that the `%` operator is the vector cross product.